Add decomposition for dynamo_export + ExportedProgram and remove None from input

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112289
* #112272
* #112271
* #112263
* __->__ #112444

This PR introduces the ability to produce GraphModules with Core ATen IR only through decompositions. It also removes `None` from user inputs as ONNX does not supports them

Tests for these features will be executed when #112289 is merged, but for reference, they are as below:

```python
    def test_log_sigmoid(self):
        # This produces op as `torch.ops.aten.log_sigmoid_forward`, instead of the more
        # conventional `torch.ops.aten.log_sigmoid`.
        class Model(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.m = torch.nn.LogSigmoid()

            def forward(self, x):
                return self.m(x)

        input = torch.randn(2)
        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            Model(), (input,), model_type=self.model_type
        )

    def test_none_input(self):
        class NoneInputModel(torch.nn.Module):
            def forward(
                self, x: torch.Tensor, y: Optional[torch.Tensor], z: torch.Tensor
            ):
                if y is None:
                    return x + z
                return x + y + z

        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            NoneInputModel(),
            (torch.randn(1, 2), None, torch.randn(1, 2)),
            model_type=self.model_type,
        )
```

Commit:
Add decomposition for dynamo_export + ExportedProgram and remove None from input

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/torch_export_graph_extractor.py:
```diff
@@ -42,6 +42,9 @@ def generate_fx(
         #     )
+        # TODO: input_adapter and output_adapter should be based on pytree flatten/unflatten API?
+
+        model = model.run_decompositions(options.decomposition_table)
```
Put it in `pre_export_passes`?

thiagocrepaldi replied:
This is tricky. Although we have implemented decomposition as a `GraphModule` pass for the `torch._dynamo.export` `fx_tracer`, `torch.export` does decomp at the `ExportedProgram` level. Through the `torch.export` + `ExportedProgram` flow, we run decompositions before we start doing passes of the `GraphModule` (using `self.pre_export_passes`)

In fact, at `generate_fx` we call `return self.pre_export_passes()` passing `original_model=model` and `fx_module=model.graph_module`. Note that we extract `GraphModule` from the `ExportedProgram.graph_module` after the decomposition was ran. Only then we start doing `GraphModule` passes (and ignore `original_model`).

In order to move the `ExportedProgram` decomposition to `pre_export_passes` we have to 
1) change `model = model.run_decompositions(options.decomposition_table)` call (which assigns a `ExportedProgram` to another ExportedProgram`) to `fx_module = original_model.run_decompositions(options.decomposition_table).graph_module` (which assigns a `GraphModule` coming from a `ExportedProgram`; that is the `generate_fx`'s role, not `pre_export_passes`)
2) Always run `fx_module: GraphModule = exported_program.run_decompositions` as the first pass and 
3) Remove `fx_module` as the input to the `pre_export_passes`, meaning `generate_fx` is not generating the `GraphModule` anymore

Although techinically possible, it mixes responsibilities of `generate_fx` and `pre_export_passes`.

BowenBao replied:
Okay. We are missing out diagnostics for decomp though.

thiagocrepaldi replied:
how should we address this?

BowenBao replied:
Need to think more on that. Good opportunity to explore how to track low level info such as these from other modules.

BowenBao commented on torch/onnx/_internal/fx/torch_export_graph_extractor.py:
```diff
@@ -42,6 +42,9 @@ def generate_fx(
         #         args=model_args,  # type: ignore[arg-type]
         #         kwargs=model_kwargs,  # type: ignore[arg-type]
         #     )
+        # TODO: input_adapter and output_adapter should be based on pytree flatten/unflatten API?
```
Unrelated outdated comment?

thiagocrepaldi replied:
nope, an actual extension of the comments above. We probably don't need to use the input adapter steps we created ro `ExportedProgram` for each change (flatten, remove none, etc). A simple inputAdfapterstep that calls the pytree to unflatten input and another to unflatten output seems more correct. The torch.export contract is based on "pytreeable" input/output

BowenBao replied:
The code today is the same as you mentioned. `FlattenInputWithTreeSpecValidationInputStep` is the pytree flatten step. If you want you can rename it to put pytree in the name. 

Or maybe you meant `ExportedProgram` needs only this step? That I don't think so since other steps are a requirement from elsewhere, e.g., onnx.

thiagocrepaldi replied:
`FlattenInputWithTreeSpecValidationInputStep  uses pytree, that is what I meant. great, no renaming needed

Commit:
Update on "Add decomposition for dynamo_export + ExportedProgram and remove None from input"


This PR introduces the ability to produce GraphModules with Core ATen IR only through decompositions. It also removes `None` from user inputs as ONNX does not supports them

Tests for these features will be executed when #112289 is merged, but for reference, they are as below:

```python
    def test_log_sigmoid(self):
        # This produces op as `torch.ops.aten.log_sigmoid_forward`, instead of the more
        # conventional `torch.ops.aten.log_sigmoid`.
        class Model(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.m = torch.nn.LogSigmoid()

            def forward(self, x):
                return self.m(x)

        input = torch.randn(2)
        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            Model(), (input,), model_type=self.model_type
        )

    def test_none_input(self):
        class NoneInputModel(torch.nn.Module):
            def forward(
                self, x: torch.Tensor, y: Optional[torch.Tensor], z: torch.Tensor
            ):
                if y is None:
                    return x + z
                return x + y + z

        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            NoneInputModel(),
            (torch.randn(1, 2), None, torch.randn(1, 2)),
            model_type=self.model_type,
        )
```



[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/torch_export_graph_extractor.py:
```diff
@@ -42,6 +42,9 @@ def generate_fx(
         #         args=model_args,  # type: ignore[arg-type]
         #         kwargs=model_kwargs,  # type: ignore[arg-type]
         #     )
+        # TODO: input_adapter and output_adapter should be based on pytree flatten/unflatten API?
```
```suggestion
```

Commit:
Update torch/onnx/_internal/fx/torch_export_graph_extractor.py

thiagocrepaldi commented:
@pytorchbot merge

thiagocrepaldi commented:
@pytorchbot merge

Commit:
Update on "Add decomposition for dynamo_export + ExportedProgram and remove None from input"


This PR introduces the ability to produce GraphModules with Core ATen IR only through decompositions. It also removes `None` from user inputs as ONNX does not supports them

Tests for these features will be executed when #112289 is merged, but for reference, they are as below:

```python
    def test_log_sigmoid(self):
        # This produces op as `torch.ops.aten.log_sigmoid_forward`, instead of the more
        # conventional `torch.ops.aten.log_sigmoid`.
        class Model(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.m = torch.nn.LogSigmoid()

            def forward(self, x):
                return self.m(x)

        input = torch.randn(2)
        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            Model(), (input,), model_type=self.model_type
        )

    def test_none_input(self):
        class NoneInputModel(torch.nn.Module):
            def forward(
                self, x: torch.Tensor, y: Optional[torch.Tensor], z: torch.Tensor
            ):
                if y is None:
                    return x + z
                return x + y + z

        self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
            NoneInputModel(),
            (torch.randn(1, 2), None, torch.randn(1, 2)),
            model_type=self.model_type,
        )
```



[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge
